src/lowering/desugar.ml

```diff
@@ -198,6 +198,9 @@ and exp' at note = function
   | S.WhileE (e1, e2) -> (whileE (exp e1) (exp e2)).it
   | S.LoopE (e1, None) -> I.LoopE (exp e1)
   | S.LoopE (e1, Some e2) -> (loopWhileE (exp e1) (exp e2)).it
+  | S.ForE (p, ({it=S.CallE (({it=S.DotE (arr, proj); _} as c0), c1, c2); _} as e1), e2)
+      when T.is_array arr.note.S.note_typ && proj.it = "vals"
+    -> (rewrite_for_to_while (pat p) arr proj c0 c1 c2 e1 (exp e2)).it
```

Comment:
I find it fishy that you are passing `c0` here when you are already passing `arr` and `proj`, as similar for other components. Is that necessary?

Maybe simply match that  `c2` is `TupE []` here. And `c1 : inst` can probably be dropped? Isn’t that always “empty” anyways?

I see you are using them in `rewrite_for_to_while` in the call to `size`. That seems a bit hackish somehow.